library/Core/src/Juvix/Core/Common/Context.hs

```diff
@@ -111,55 +128,64 @@ topList T {topLevelMap} = HashMap.toList topLevelMap
 -- This function also keeps the invariant that there is only one CurrentNameSpace
 -- tag
 switchNameSpace ::
-  NameSymbol.T -> T term ty sumRep -> Either PathError (T term ty sumRep)
+  NameSymbol.T -> T term ty sumRep -> IO (Either PathError (T term ty sumRep))
 switchNameSpace newNameSpace t@T {currentName}
   | removeTopName newNameSpace == removeTopName currentName =
-    Lib.Right t
-  | otherwise =
+    pure (Lib.Right t)
+  | otherwise = do
     let addCurrentName t startingContents newCurrName =
           (addGlobal' t)
             { currentName = removeTopName newCurrName,
               currentNameSpace = startingContents
             }
+        -- addGlobal' adds the current name space back to the environment map
         addGlobal' t@T {currentNameSpace} =
           -- we have to add top to it, or else if it's a single symbol, then
           -- it'll be added to itself, which is bad!
           addGlobal
             (addTopNameToSngle currentName)
-            (Record currentNameSpace Nothing)
+            (Record currentNameSpace)
             t
         addCurrent = addGlobal newNameSpace CurrentNameSpace
         qualifyName =
           currentName <> newNameSpace
-     in case addPathWithValue newNameSpace CurrentNameSpace t of
-          Lib.Right t ->
-            -- check if we have to qualify Name
-            case t !? newNameSpace of
-              Just (Current _) ->
-                Lib.Right (addCurrentName t NameSpace.empty qualifyName)
-              _ ->
-                Lib.Right (addCurrentName t NameSpace.empty newNameSpace)
-          -- the namespace may already exist
-          Lib.Left er ->
-            -- TODO ∷ refactor with resolveName
-            -- how do we add the namespace back to the private area!?
-            case t !? newNameSpace of
-              Just (Current (NameSpace.Pub (Record def _))) ->
-                Lib.Right (addCurrent (addCurrentName t def qualifyName))
-              Just (Current (NameSpace.Priv (Record def _))) ->
-                Lib.Right (addCurrent (addCurrentName t def qualifyName))
-              Just (Outside (Record def _))
-                -- figure out if we contain what we are looking for!
-                | NameSymbol.prefixOf (removeTopName newNameSpace) currentName ->
-                  -- if so update it!
-                  case addGlobal' t !? newNameSpace of
-                    Just (Outside (Record def _)) ->
-                      Lib.Right (addCurrent (addCurrentName t def newNameSpace))
-                    _ -> error "doesn't happen"
-                | otherwise ->
-                  Lib.Right (addCurrent (addCurrentName t def newNameSpace))
-              Nothing -> Lib.Left er
-              Just __ -> Lib.Left er
+    ret <- addPathWithValue newNameSpace CurrentNameSpace t
+    case ret of
```

Comment:
this needs way more comments, reviewing this I am pretty confused as to what all the variables are

try adding a comment for each branch of each case switch, to start

Reply:
thank you, the new comments are much better